README.md
[![Travis-CI Build Status](https://travis-ci.org/ropensci/onekp.svg?branch=master)](https://travis-ci.org/ropensci/onekp)
[![Coverage Status](https://img.shields.io/codecov/c/github/ropensci/onekp/master.svg)](https://codecov.io/github/ropensci/onekp?branch=master)

# onekp

The [1000 Plants initiative
(1KP)](https://sites.google.com/a/ualberta.ca/onekp/) provides the
transcriptome sequences to over 1000 plants from diverse lineages. `onekp`
allows researchers in plant genomics and transcriptomics to access this dataset
through a simple R interface. The metadata for each transcriptome project is
scraped from the 1KP project website. This metadata includes the species,
tissue, and research group for each sequence sample. `onekp` leverages the
taxonomy program `taxizedb`, a local database version of `taxize` package, to
allow filtering of the metadata by taxonomic group (entered as either a taxon
name or NCBI ID). The raw nucleotide or translated peptide sequence can then be
downloaded for the full, or filtered, table of transcriptome projects. 

## Alternatives to `onekp`

The data may also be accessed directly through CyVerse (previously iPlant).
CyVerse efficiently distributes data using the iRODS data system. This approach
is preferable for high-throughput cases or in where iRODS is already in play.
Further, accessing data straight from the source at CyVerse is more stable than
scraping it from project website. However, the `onekp` R package is generally
easier to use (no iRODS dependency or CyVerse API) and offers powerful
filtering solutions. 

## Contact info

1KP staff

 * [Gane Ka-Shu Wong](https://sites.google.com/a/ualberta.ca/professor-gane-ka-shu-wong/) - Principal investigator

 * [Michael Deyholos](mkdeyholos@gmail.com) - Alberta co-investigator

 * [Yong Zhang](zhangy@genomics.org.cn) - Shenzhen co-investigator

 * [Eric Carpenter](ejc@ualberta.ca) - Database manager

R package maintainer

 * [Zebulun Arendsee](arendsee@iastate.edu)


## Installation

`onekp` is on CRAN, but currently is a little out of date. So for now it is
better to install through github. 


```r
library(devtools)
install_github('ropensci/onekp')
```

## Examples

Retrieve the protein and gene transcript FASTA files for two 1KP transcriptomes: 


```r
onekp <- retrieve_onekp()
seqs <- filter_by_code(onekp, c('URDJ', 'ROAP'))
download_peptides(seqs, 'oneKP/pep')
download_nucleotides(seqs, 'oneKP/nuc')
```

This will create the following directory:

```
oneKP
 ├── nuc 
 │   ├── ROAP.fna
 │   └── URDJ.fna
 └── pep
     ├── ROAP.faa
     └── URDJ.faa
```

`onekp` can also filter by species names, taxon ids, or clade.


```r
# filter by species name
filter_by_species(onekp, 'Pinus radiata')

# filter by species NCBI taxon ID
filter_by_species(onekp, 3347)

# filter by clade name scientific name (get all data for the Brassicaceae family)
filter_by_clade(onekp, 'Brassicaceae')

# filter by clade NCBI taxon ID
filter_by_clade(onekp, 3700)
```

So to get the protein sequences for all species in Brassicaceae:


```r
onekp <- retrieve_onekp()
seqs <- filter_by_clade(onekp, 'Brassicaceae')
download_peptides(seqs, 'oneKP/pep')
download_nucleotides(seqs, 'oneKP/nuc')
```

# Funding

Development of this R package was supported by the National Science Foundation under Grant No. IOS 1546858.

# Contributing

We welcome any contributions!

 * By participating in this project you agree to abide by the terms outlined in
   the [Contributor Code of Conduct](CONDUCT.md).

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
# onekp 0.2.1

 First rOpenSci version

 * Transfer ownership to rOpenSci

 * Update README
 
 * Add .github files

 * Add reviewers to DESCRIPTION

# onekp 0.2.0

 The goal of this version is to address all issues raised by the rOpenSci
 editor (Scott Chamberlain) and my two reviewers: Jessica Minnier and Zachary
 Foster.

## Updates based on editor comments

 * Added `URL` tag to DESCRIPTION

 * Added `BugReports` tag to DESCRIPTION

 * Wrapped all lines longer than 80 characters

 * Removed unused imports `readr` and `tidyr`

 * Split test functions into smaller test files

 * Add `skip_on_cran` and `skip_on_travis` to `retrieve_onekp` test

 * Save metadata as package data: `data(onekp)` 

 * Add minimum version to `taxizedb`

## Updates based on Zachary Foster's review

 * Add space between equal signs in argument lists

 * Change package name from `oneKP` to `onekp`

 * Let user set the download directory (defaulting to a tempdir).

 * Re-export the `magrittr` `%>%` operator

 * Update README

 * Add CONDUCT.md

 * Add R `devel` test to travis

 * More package-level documentation

 * Do not export print.OneKP function

 * Add documentation details to `retrieve_onekp`

 * Update intro vignette

 * Do not run examples that access the internet (call `retrieve_onekp`)

## Updates based on Jessica Minnier's review  

 There was some overlap between the two reviews, here I just list updates made that are unique to this review.

 * Clarify `filter_by_code` and `filter_by_species` examples

 * Cleanup the `download` documentation

 * Add `absolute` parameter to `download_*` functions. If TRUE, this returns
   the absolute path to the retrieved data.

 * Add handling for downloading nothing: if everything is filtered out, just
   return an empty character vector.

 * Add test coverage for corner cases

   - filtering by clade with a clade that is not in the NCBI common tree raises
     an error.

   - filtering with valid input, but where no rows are selected, returns
     a OneKP object with 0 rows (no error).

   - download an empty object returns an empty vector of filenames (no error)

# onekp 0.1.0

 * Initial release


R/filter.R
#' Filter a OneKP object
#'
#' @param x OneKP object
#' @param code character vector of 1KP IDs (e.g. URDJ)
#' @param clade vector of clade-level NCBI taxonomy IDs or scientific names
#' @param species vector of species-level scientific names or NCBI taxonomy IDs
#' @return OneKP object
#' @name filter
#' @examples
#' data(onekp)
#'
#' # filter by 1KP ID
#' filter_by_code(onekp, c('URDJ', 'ROAP'))
#'
#' # filter by species name
#' filter_by_species(onekp, 'Pinus radiata')
#'
#' # filter by species NCBI taxon ID
#' filter_by_species(onekp, 3347)
#'
#' # filter by clade name scientific name
#' filter_by_clade(onekp, 'Brassicaceae')
#'
#' # filter by clade NCBI taxon ID
#' filter_by_clade(onekp, 3700)
NULL

#' @rdname filter
#' @export
filter_by_code <- function(x, code) {
  x@table <- x@table[x@table[['code']] %in% code, ]
  x
}

#' @rdname filter
#' @export
filter_by_clade <- function(x, clade) {
  if(all(is.character(clade))){
    taxids <- taxizedb::name2taxid(clade) 
    if(any(is.na(taxids))){
      msg <- "The following clades are not in NCBI: "
      stop(msg, paste(clade[is.na(taxids)], collapse=", "))
    }
    clade <- taxids
  }

  # Get all NCBI taxonomy IDs descending from the input clades
  taxa <- lapply(taxizedb::downstream(
    clade,
    downto            = 'species',
    ambiguous_node    = TRUE,
    ambiguous_species = TRUE
  ), function(d){
   d$childtaxa_id 
  }) %>% unlist %>% unname

  # Find the common ids
  onekp_species <- intersect(taxa, x@table$tax_id)

  filter_by_species(x, onekp_species)
}

#' @rdname filter
#' @export
filter_by_species <- function(x, species) {
  selection <- if(all(grepl('^[0-9]+$', species, perl = TRUE))){
    x@table$tax_id %in% species
  } else {
    species <- gsub('_', ' ', species)
    x@table$species %in% species
  }
  x@table <- x@table[selection, ]
  x
}
